src/growable.c

```diff
@@ -0,0 +1,274 @@
+  UNPROTECT(1);
+  return ret;
+}
+
+R_xlen_t growable_max_size(SEXP x) {
+  return TRUELENGTH(x);
+}
+
+void growable_resize(SEXP x, R_xlen_t newsize) {
+  R_xlen_t max_size;
+  if (newsize > (max_size = growable_max_size(x))) internal_error(
+    __func__, "newsize=%g > max_size=%g",
+    (double)newsize, (double)max_size
+  );
+  SETLENGTH(x, newsize);
+}
+
+Rboolean is_growable(SEXP x) {
+  return isVector(x) && TRUELENGTH(x) >= XLENGTH(x)
+#if R_VERSION >= R_Version(3, 4, 0)
+    && IS_GROWABLE(x)
+#endif
+  ;
+}
+
+// Assuming no ALTREP columns
+SEXP make_growable(SEXP x) {
+  if (TRUELENGTH(x) < XLENGTH(x)) SET_TRUELENGTH(x, XLENGTH(x));
+  SET_GROWABLE_BIT(x);
+  return x;
+}
+
+#else
+
+#include <R_ext/Altrep.h>
+
+static R_altrep_class_t dta_grow_string, dta_grow_integer, dta_grow_logical, dta_grow_real, dta_grow_complex, dta_grow_raw, dta_grow_list;
+static Rcomplex NA_COMPLEX = { 0, };
+
+/*
+ALTREP class layout:
+data1 = underlying vector
+data2 = its current length stored as a length-1 REALSXP
+Unless we implement an Unserialize method, this can be changed any time.
+Classes have been released on CRAN with a Serialized_state/Unserialize pair will have to stay as they have been defined in order to keep *.rds files readable.
+*/
+
+static R_xlen_t altall_Length(SEXP x) {
+  return (R_xlen_t)REAL(R_altrep_data2(x))[0];
+}
+
+#define make_inspect_method(classname) \
+  static Rboolean alt##classname##_Inspect( \
+    SEXP x, int pre, int deep, int pvec, \
+    void (*inspect_subtree)(SEXP x, int pre, int deep, int pvec) \
+  ) { \
+    (void)pre; (void)deep; (void)pvec; (void)inspect_subtree; \
+    Rprintf("data.table::growable" #classname "_v0(truelength=%g) ", (double)XLENGTH(R_altrep_data1(x))); \
+    return FALSE; \
+  }
+make_inspect_method(string)
+make_inspect_method(integer)
+make_inspect_method(logical)
+make_inspect_method(real)
+make_inspect_method(complex)
+make_inspect_method(raw)
+make_inspect_method(list)
+#undef make_inspect_method
+
+#define make_dataptr_method(class, accessor) \
+  static void * alt##class##_Dataptr(SEXP x, Rboolean writable) { \
+    (void)writable; \
+    return (void*)accessor(R_altrep_data1(x)); \
+  }
+make_dataptr_method(string, STRING_PTR_RO)
+make_dataptr_method(integer, INTEGER)
+make_dataptr_method(logical, LOGICAL)
+make_dataptr_method(real, REAL)
+make_dataptr_method(complex, COMPLEX)
+make_dataptr_method(raw, RAW)
+make_dataptr_method(list, DATAPTR_RO) // VECTOR_PTR_RO to appear in R-4.5
+#undef make_dataptr_method
+
+static const void * altall_Dataptr_or_null(SEXP x) { return DATAPTR_RO(x); }
+
+// lots of boilerplate, but R calling *_ELT one by one would be far too slow
+#define make_extract_subset_method(class, type, accessor, NA) \
+  static SEXP alt##class##_Extract_subset(SEXP x, SEXP indx, SEXP call) { \
+    (void)call; \
+    indx = PROTECT(coerceVector(indx, REALSXP)); \
+    double * ii = REAL(indx); \
+    R_xlen_t rlen = XLENGTH(indx), mylen = XLENGTH(x); \
```

Comment:
maybe I'm misunderstanding extract_subset but shouldn't this only be valid for the actual size and not the full overallocated length?

Hence, `mylen = (R_xlen_t)REAL(R_altrep_data2(x))[0];` ?

Reply:
`XLENGTH(x)` goes through the ALTREP dispatch and lands in `altall_Length`, but we could definitely save a function call and make the code clearer.